extensions/HyperVExtension/src/HyperVExtension/Exceptions/GuestOsVersionException.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Globalization;
+using System.Management.Automation;
+using System.Text;
+using HyperVExtension.Common;
+using HyperVExtension.Helpers;
+
+namespace HyperVExtension.Exceptions;
+
+internal class GuestOsVersionException : Exception
+{
+    public GuestOsVersionException(IStringResource stringResource, Exception innerException, Dictionary<string, string>? guestOsProperties)
+        : base(stringResource.GetLocalized("FailedToDetermineGuestOsVersion"), innerException)
+    {
+        HResult = innerException.HResult;
+        GuestOsProperties = guestOsProperties;
+    }
+
+    protected GuestOsVersionException(string message, Dictionary<string, string>? guestOsProperties)
+        : base(message)
+    {
+        GuestOsProperties = guestOsProperties;
+    }
+
+    public Dictionary<string, string>? GuestOsProperties { get; }
+
+    public override string ToString()
+    {
+        StringBuilder message = new();
+        message.Append(CultureInfo.InvariantCulture, $"{Message}.");
+        if (InnerException != null)
+        {
+            message.Append(CultureInfo.InvariantCulture, $" {InnerException}.");
+        }
+
+        if (GuestOsProperties != null)
+        {
+            GuestOsProperties.TryGetValue(HyperVStrings.OSPlatformId, out var osPlatformId);
+            GuestOsProperties.TryGetValue(HyperVStrings.OSVersion, out var osVersion);
+            GuestOsProperties.TryGetValue(HyperVStrings.OSName, out var osName);
+            if ((osPlatformId != null) ||

```

Comment:
Not blocking but I feel like maybe it would be good reporting info for us to know if one was null or not in the strings below. As in, should we capture this null info in the strings instead of keeping them behind the null checks?

Reply:
I thought about that and decided not to output "null", "no data", or whatever so users don't get the impression that it a mistake in the code. The fact that one of them is not in the output means it was null.